pkg/agent/openflow/packetin.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package openflow
+
+import (
+	"time"
+
+	"github.com/contiv/ofnet/ofctrl"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+)
+
+type ofpPacketInReason uint
+
+type PacketInHandler interface {
+	HandlePacketIn(pktIn *ofctrl.PacketIn) error
+	GetHandlerName() string
+}
+
+const (
+	// Action explicitly output to controller.
+	ofprAction ofpPacketInReason = 1
+)
+
+func (c *client) RegisterPacketInHandler(packetInHandler interface{}) {
+	handler, ok := packetInHandler.(PacketInHandler)
+	if !ok {
+		klog.Errorf("Invalid Traceflow controller.")
+		return
+	}
+	c.packetInHandlers[handler.GetHandlerName()] = handler
+}
+
+func (c *client) StartPacketInHandler(stopCh <-chan struct{}) {
+	if len(c.packetInHandlers) == 0 {
```

Comment:
Again, there is a race condition for `c.packetInHandlers`
`RegisterPacketInHandler` and `StartPacketInHandler` runs in two goroutines concurrently, there is no guarantee of their order. More seriously, the program might panic if unlucky.
I would strongly suggest not starting traceflow introduced goroutines/modules if the feature is not enabled given that race conditions I have observed.

Reply:
Move c.ofClient.RegisterPacketInHandler to NewTraceflowController to guarantee the sequence.

Reply:
We need more code review and test cases to discover all potential race conditions.
@tnqn , @gran-vmv , could we design some E2E cases to discover those issues? If we run multiple traceflow requests in parallel, this stress test can trigger unexpected concurrency problems?

Reply:
Agreed we should not StartPacketInHandler and start traceflow controller when TF feature is not enabled - I do not see why we do that?

Also I think we could just StartPacketInHandler() in Traceflow controller after the registration is done, at least before we have another packetIn handler.

Reply:
Fixed.
Moved below code snippet from "go traceflowController.Run(stopCh)" to NewTraceflowController(). Then we can guarantee the sequence.
```
	// Register packetInHandler
	c.ofClient.RegisterPacketInHandler("traceflow", c)
```